src/main/kotlin/org/rust/ide/folding/RustFoldingBuilder.kt

```diff
@@ -0,0 +1,85 @@
+package org.rust.ide.folding
+
+import com.intellij.lang.ASTNode
+import com.intellij.lang.folding.FoldingBuilderEx
+import com.intellij.lang.folding.FoldingDescriptor
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.PsiTreeUtil
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.impl.RustFile
+import java.util.*
+
+class RustFoldingBuilder() : FoldingBuilderEx() {
+    override fun getPlaceholderText(node: ASTNode): String? {
+        return "{...}"
+    }
+
+    override fun buildFoldRegions(root: PsiElement, document: Document, quick: Boolean): Array<out FoldingDescriptor> {
+        if (root !is RustFile) return emptyArray()
+
+        val descriptors: MutableList<FoldingDescriptor> = ArrayList()
+
+        PsiTreeUtil.findChildrenOfType(root, RustBlockExpr::class.java).forEach {
```

Comment:
Can you add

```Kotlin
inline fun <reified T : PsiElement> PsiElement.descendantsOfType(): Collection<T>
```

to [RustPsiExtensions]? 

Alternatively, I think this can be expressed shorter (and perhaps faster) if you use recursive `RustVisitor`. 

[RustPsiExtensions]: https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/psi/util/RustPsiExtensions.kt#L25